File: express-starter/src/routes/items.ts
import { Router } from 'express';
import { z } from 'zod';
import { createError } from '../middleware/error.js';

const router = Router();

// In-memory storage for demo
const items = new Map<number, Item>();
let counter = 0;

// Schemas
const createItemSchema = z.object({
  name: z.string().min(1).max(100),
  description: z.string().max(500).optional(),
  price: z.number().positive(),
  isActive: z.boolean().default(true),
});

const updateItemSchema = createItemSchema.partial();

const querySchema = z.object({
  page: z.coerce.number().int().positive().default(1),
  pageSize: z.coerce.number().int().min(1).max(100).default(10),
  activeOnly: z.coerce.boolean().default(false),
});

interface Item {
  id: number;
  name: string;
  description?: string;
  price: number;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

// List items
router.get('/', (req, res) => {
  const query = querySchema.parse(req.query);

  let itemList = Array.from(items.values());

  if (query.activeOnly) {
    itemList = itemList.filter((item) => item.isActive);
  }

  const total = itemList.length;
  const totalPages = Math.ceil(total / query.pageSize);
  const start = (query.page - 1) * query.pageSize;
  const paginatedItems = itemList.slice(start, start + query.pageSize);

  res.json({
    items: paginatedItems,
    total,
    page: query.page,
    pageSize: query.pageSize,
    totalPages,
  });
});

// Create item
router.post('/', (req, res) => {
  const data = createItemSchema.parse(req.body);

  counter += 1;
  const now = new Date();

  const item: Item = {
    id: counter,
    name: data.name,
    description: data.description,
    price: data.price,
    isActive: data.isActive,
    createdAt: now,
    updatedAt: now,
  };

  items.set(counter, item);

  res.status(201).json(item);
});

// Get item
router.get('/:id', (req, res, next) => {
  const id = parseInt(req.params.id, 10);
  const item = items.get(id);

  if (!item) {
    return next(createError('Item not found', 404));
  }

  res.json(item);
});

// Update item
router.patch('/:id', (req, res, next) => {
  const id = parseInt(req.params.id, 10);
  const item = items.get(id);

  if (!item) {
    return next(createError('Item not found', 404));
  }

  const updates = updateItemSchema.parse(req.body);

  const updatedItem: Item = {
    ...item,
    ...updates,
    updatedAt: new Date(),
  };

  items.set(id, updatedItem);

  res.json(updatedItem);
});

// Delete item
router.delete('/:id', (req, res, next) => {
  const id = parseInt(req.params.id, 10);

  if (!items.has(id)) {
    return next(createError('Item not found', 404));
  }

  items.delete(id);
  res.status(204).send();
});

export default router;

File: fastapi-starter/app/api/items.py
"""Example items API endpoints."""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel, Field

router = APIRouter()


# In-memory storage for demo
_items_db: dict[int, dict] = {}
_counter = 0


class ItemCreate(BaseModel):
    """Schema for creating an item."""

    name: str = Field(..., min_length=1, max_length=100, examples=["My Item"])
    description: str | None = Field(None, max_length=500)
    price: float = Field(..., gt=0, examples=[29.99])
    is_active: bool = Field(default=True)


class ItemUpdate(BaseModel):
    """Schema for updating an item."""

    name: str | None = Field(None, min_length=1, max_length=100)
    description: str | None = Field(None, max_length=500)
    price: float | None = Field(None, gt=0)
    is_active: bool | None = None


class ItemResponse(BaseModel):
    """Schema for item response."""

    id: int
    name: str
    description: str | None
    price: float
    is_active: bool


class PaginatedResponse(BaseModel):
    """Schema for paginated response."""

    items: list[ItemResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


@router.get("/items", response_model=PaginatedResponse)
async def list_items(
    page: int = Query(1, ge=1, description="Page number"),
    page_size: int = Query(10, ge=1, le=100, description="Items per page"),
    active_only: bool = Query(False, description="Filter active items only"),
):
    """List all items with pagination."""
    items = list(_items_db.values())

    if active_only:
        items = [i for i in items if i["is_active"]]

    total = len(items)
    total_pages = (total + page_size - 1) // page_size

    start = (page - 1) * page_size
    end = start + page_size
    paginated_items = items[start:end]

    return {
        "items": paginated_items,
        "total": total,
        "page": page,
        "page_size": page_size,
        "total_pages": total_pages,
    }


@router.post("/items", response_model=ItemResponse, status_code=201)
async def create_item(item: ItemCreate):
    """Create a new item."""
    global _counter
    _counter += 1

    new_item = {
        "id": _counter,
        "name": item.name,
        "description": item.description,
        "price": item.price,
        "is_active": item.is_active,
    }
    _items_db[_counter] = new_item

    return new_item


@router.get("/items/{item_id}", response_model=ItemResponse)
async def get_item(item_id: int):
    """Get a specific item by ID."""
    if item_id not in _items_db:
        raise HTTPException(status_code=404, detail="Item not found")

    return _items_db[item_id]


@router.patch("/items/{item_id}", response_model=ItemResponse)
async def update_item(item_id: int, item: ItemUpdate):
    """Update an existing item."""
    if item_id not in _items_db:
        raise HTTPException(status_code=404, detail="Item not found")

    stored_item = _items_db[item_id]
    update_data = item.model_dump(exclude_unset=True)

    for field, value in update_data.items():
        stored_item[field] = value

    return stored_item


@router.delete("/items/{item_id}", status_code=204)
async def delete_item(item_id: int):
    """Delete an item."""
    if item_id not in _items_db:
        raise HTTPException(status_code=404, detail="Item not found")

    del _items_db[item_id]
